tracing-appender/src/rolling.rs

```diff
@@ -327,17 +334,155 @@ impl Rotation {
+#[derive(Debug, Clone, PartialEq)]
+pub struct Prefixed {
+    log_directory: PathBuf,
+    prefix: OsString,
+}
+
+impl Prefixed {
+    /// Create a new [`Prefixed`] `FileTemplate`.
+    pub fn new<D, P>(log_directory: D, prefix: P) -> Self
+    where
+        D: AsRef<Path>,
+        P: AsRef<OsStr>,
+    {
+        Prefixed {
+            log_directory: log_directory.as_ref().to_owned(),
+            prefix: prefix.as_ref().to_owned(),
         }
     }
 }
 
+impl FilenameTemplate for Prefixed {
+    fn next_log_file(&mut self, date: &DateTime<Utc>, rotation: &Rotation) -> PathBuf {
+        let mut last_segment = self.prefix.clone();
+
+        if let Some(timestamp) = format_date_for_rotation(date, rotation) {
+            last_segment.push(".");
+            last_segment.push(&timestamp);
+        }
+
+        self.log_directory.join(last_segment)
+    }
+}
+
+/// Create a new [`FilenameTemplate`] which uses a common file name and
+/// extension.
+///
+/// # Examples
+///
+/// ```rust
+/// # use chrono::prelude::*;
+/// # use tracing_appender::rolling::{Rotation, FilenameTemplate};
+/// # use std::path::Path;
+/// let mut template = tracing_appender::rolling::with_name_and_extension(
+///     "/var/log",
+///     "MyApplication",
+///     "log",
+/// );
+///
+/// let date = Utc.ymd(2020, 2, 1).and_hms(12, 20, 15);
+/// let rotation = Rotation::DAILY;
+///
+/// let filename = template.next_log_file(&date, &rotation);
+/// assert_eq!(filename, Path::new("/var/log/MyApplication.2020-02-01.log"));
+/// ```
+pub fn with_name_and_extension<D, N, E>(
+    log_directory: D,
+    name: N,
+    extension: E,
+) -> WithNameAndExtension
+where
+    D: AsRef<Path>,
+    N: AsRef<OsStr>,
+    E: AsRef<OsStr>,
+{
+    WithNameAndExtension::new(log_directory, name, extension)
+}
+
+/// A [`FilenameTemplate`] with a timestamp appended to a common prefix.
+#[derive(Debug, Clone, PartialEq)]
+pub struct WithNameAndExtension {
+    log_directory: PathBuf,
+    name: OsString,
+    extension: OsString,
+}
+
+impl WithNameAndExtension {
+    /// Create a new [`Prefixed`] `FileTemplate`.
+    pub fn new<D, N, E>(log_directory: D, name: N, extension: E) -> Self
+    where
+        D: AsRef<Path>,
+        N: AsRef<OsStr>,
+        E: AsRef<OsStr>,
+    {
+        WithNameAndExtension {
+            log_directory: log_directory.as_ref().to_owned(),
+            name: name.as_ref().to_owned(),
+            extension: extension.as_ref().to_owned(),
+        }
+    }
+}
+
+impl FilenameTemplate for WithNameAndExtension {
+    fn next_log_file(&mut self, date: &DateTime<Utc>, rotation: &Rotation) -> PathBuf {
+        let mut last_segment = self.name.clone();
+
+        if let Some(timestamp) = format_date_for_rotation(date, rotation) {
+            last_segment.push(".");
+            last_segment.push(&timestamp);
+        }
+
+        last_segment.push(".");
+        last_segment.push(&self.extension);
+
+        self.log_directory.join(last_segment)
```

Comment:
nit: We could have the `WithNameAndExtension` wrap `Prefixed` and simply append the extension to the end of it, hence reducing some code duplicaton in `next_log_file`.

Either approach is fine with me though.

Reply:
I had a play around with this approach and it doesn't look like we'll save much. 

You can only manipulate a `PathBuf` at the element level and can't append to the last segment directly. That'd mean we would call `Prefixed::next_log_file()` then copy its current extension (the timestamp) to a buffer, append the desired extension (e.g. `log`), then use `PathBuf::set_extension()` to swap in the buffer's contents. 

Alternatively we could pull the path construction logic out into some helper function that accepts a `&[&OsStr]` (e.g. `fn format_filename(directory: &OsStr, filename_bits: &[&OsStr]) -> PathBuf`.

 Both approaches feel like they'd make the code less readable and add unnecessary indirection/cognitive load.